core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -1689,6 +1690,123 @@ RexNode simplifyAnd2ForUnknownAsFalse(List<RexNode> terms,
     }
   }
 
+  /**
+   * <p>Simplifies AND/OR condition that has common expressions,
+   * extract and eliminate/merge them as much as possible.</p>
+   * <ul>
+   * <li>(a OR b) AND (a OR b OR c OR d) =&gt; (a OR b)
+   * <li>(a OR b OR c OR d) AND (a OR b OR e OR f) =&gt; (a OR b) OR ((c OR d) AND (e OR f))
+
+   * <li>(a AND b) OR (a AND b AND c) =&gt; (a AND b)
+   * <li>(a AND b AND c AND d) OR (a AND b AND e AND f) =&gt;
+   * (a AND b) AND ((c AND d) OR (e AND f))
+   * </ul>
+   * <p> The difference between {@link #simplifyAnd} is that {@link #simplifyAnd} mainly
+   * simplifies expressions whose answer can be determined without evaluating both sides,
+   * like: FALSE AND (xxx) =&gt; FALSE</p>
+   */
+  public RexNode eliminateCommonExprInCondition(RexNode call) {
```

Comment:
This method seems to be never used, so it is not clear how should it interact with other simplify methods.

Reply:
See CALCITE-4375, I split sub-task for this JIRA, you can see the comment I reply Julian, "and later we can think to add this logic to RexSimplify's main process or just used to simplify the join condition."

IMO, we can first use it to  simplify the join condition, latter we can add this to RexSimplify's main process like Spark.

Reply:
Just pasted from CALCITE-4375 if you have no time to read:
SQL:
```
SELECT * FROM emps,depts
WHERE
(emps.name = depts.name AND empno=1)
OR
(emps.name = depts.name AND empno=2)
```

And the join after optimizer is:

`EnumerableNestedLoopJoin(condition=[OR(AND(=($1, $11), =($0, 1)), AND(=($1, $11), =($0, 2)))], joinType=[inner])`


In fact ($1, $11) can be extracted, and the join can be:

`HashJoin(condition=[AND(=($1, $11), OR(=($0, 1), =($0, 2)))], joinType=[inner]) `

and then the or cond can be pushed down, and the SQL is like

`select * from emps, depts on( emps.name = depts.name ) where (empno=1 OR empno=2)`

Reply:
@Aaaaaaron if I understand correctly the idea of this PR is just pushing this new method (unused for the moment).
Then, in a separate PR, plug this method to the rule (I guess JoinPushExpressionsRule?).
Finally, to be discussed if this method should / should not be applied to RexSimplify.
Is that correct?

Reply:
@rubenada Yes, I think this improvement can be a common optimize, so I open a sub-task for this, if you see my origin JIRA, you can see the main point is in join optimize, I don't want this optimize only attached to this.

I think the optimize is a kind of simplify, maybe we can merge to RexSimplify's main process, like Spark: org.apache.spark.sql.catalyst.optimizer.BooleanSimplification, hope to hear your opinion.

Reply:
I uderstand  @Aaaaaaron , thanks for the explanation.
IMHO the current PR as it is (a new method that is not used, apart from the unit test), does not make much sense. I would prefer to see a PR combining 1 & 2 from my previous comment (i.e. this new method, used by a certain join rule, so that we can see it "in action").
Having said that, if people think otherwise, I would not oppose to merging the current PR as it is.

Reply:
Sure, I will combine 1&2.

Reply:
@rubenada I've updated my PR, could you take a look?